polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -298,6 +324,154 @@ async fn run<Context>(
+}
+
+// Returns true if the node is an authority in the next session.
+async fn check_next_session_authority<Sender>(
+	sender: &mut Sender,
+	keystore: KeystorePtr,
+	relay_parent: Hash,
+) -> bool
+where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	// In spite of function name here we request past, present and future authorities.
+	// It's ok to stil prepare PVFs in other cases, but better to request only future ones.
+	let Ok(Ok(authorities)) = util::request_authorities(relay_parent, sender).await.await else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch authorities from runtime API",
+		);
+		return false
+	};
+
+	authorities
+		.iter()
+		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]))
+}
+
+// Sends PVF with unknown code hashes to the validation host returning the list of code hashes sent.
+async fn prepare_pvfs_for_current_candidates<Sender>(
+	sender: &mut Sender,
+	mut validation_backend: impl ValidationBackend,
+	relay_parent: Hash,
+	already_prepared: &HashSet<ValidationCodeHash>,
+) -> Option<Vec<ValidationCodeHash>>
+where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	let Ok(Ok(events)) = util::request_candidate_events(relay_parent, sender).await.await else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch candidate events from runtime API",
+		);
+		return None
+	};
+	let code_hashes = events
+		.into_iter()
+		.filter_map(|e| match e {
+			CandidateEvent::CandidateBacked(receipt, ..) => {
+				let h = receipt.descriptor.validation_code_hash;
+				if already_prepared.contains(&h) {
+					None
+				} else {
+					Some(h)
+				}
+			},
+			_ => None,
+		})
+		.collect::<Vec<_>>();
+
+	let Ok(executor_params) = util::executor_params_at_relay_parent(relay_parent, sender).await
+	else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch executor params for the session",
+		);
+		return None
+	};
+	let timeout = pvf_prep_timeout(&executor_params, PvfPrepKind::Precheck);
+
+	let mut active_pvfs = vec![];
+	let mut processed_code_hashes = vec![];
+	for code_hash in code_hashes {
+		let Ok(Ok(Some(validation_code))) =
+			util::request_validation_code_by_hash(relay_parent, code_hash, sender)
+				.await
+				.await
+		else {
+			gum::warn!(
+				target: LOG_TARGET,
+				?relay_parent,
+				?code_hash,
+				"cannot fetch validation code hash from runtime API",
+			);
+			continue;
+		};
+
+		let pvf = match sp_maybe_compressed_blob::decompress(
+			&validation_code.0,
+			VALIDATION_CODE_BOMB_LIMIT,
+		) {
+			Ok(code) => PvfPrepData::from_code(
+				code.into_owned(),
+				executor_params.clone(),
+				timeout,
+				PrepareJobKind::Prechecking,
+			),
+			Err(e) => {
+				gum::debug!(target: LOG_TARGET, err=?e, "cannot decompress validation code");
+				continue
+			},
+		};
+
+		active_pvfs.push(pvf);
+		processed_code_hashes.push(code_hash);
+	}
+
+	if let Err(err) = validation_backend.heads_up(active_pvfs).await {
```

Comment:
Considering `HOST_MESSAGE_QUEUE_SIZE` is 10 can this block ?

Reply:
It could be a problem if we stall the main loop here while we are active validator. Processing of any incoming validation requests will also be delayed by up to 100 (n_cores) validation code decompressions.

Reply:
ARAIR the whole subsystem has been moved to the blocking pool (#3122)
Yes, it would indeed be good to offload those decompressions to a separate task on a blocking pool; we discussed that. At that point, it seemed like firing cannons at sparrows. But in the context of #5012 it makes much more sense now.
Still, that's out of the scope of this issue. Raised #5071.

Reply:
Returning to the original question. We don't do any work when we're an active validator. In other cases we allow to prepare only one PVF per block to not load the validator. It's plenty of time to prepare all PVFs during the session if a validator connected in the beginning but don't overload if it connected in the end.